Tests/DataPipeline/DataPipelineCompatibilityTests.swift

```diff
@@ -0,0 +1,438 @@
+        let filteredEvents = allEvents.filter { $0.eventType == "track" }
+        XCTAssertEqual(filteredEvents.count, 1, "too many events received")
+
+        guard let savedEvent = filteredEvents.last else {
+            XCTFail("saved event must not be nil")
+            return
+        }
+
+        XCTAssertEqual(savedEvent[keyPath: "userId"] as? String, givenIdentifier)
+    }
+
+    // MARK: device
+
+    func test_registerDeviceToken_expectFinalJSONHasCorrectKeysAndValues() {
+        let givenIdentifier = String.random
+        let givenToken = String.random
+        let expectedData = deviceInfoMock.configureWithMockData()
+
+        customerIO.identify(identifier: givenIdentifier)
+        customerIO.registerDeviceToken(givenToken)
+
+        let allEvents = readTypeFromStorage(key: Storage.Constants.events)
+        let filteredEvents = allEvents.filter { $0.eventType == "track" }
+        XCTAssertEqual(filteredEvents.count, 1, "too many events received")
+
+        guard let savedEvent = filteredEvents.first else {
+            XCTFail("saved event must not be nil")
+            return
+        }
+
+        XCTAssertEqual(savedEvent[keyPath: "event"] as? String, "Device Created or Updated")
+        XCTAssertEqual(savedEvent[keyPath: "userId"] as? String, givenIdentifier)
+        XCTAssertEqual(savedEvent[keyPath: "context.device.token"] as? String, givenToken)
+        XCTAssertFalse(savedEvent.containsKey("last_used"))
+        XCTAssertFalse(savedEvent.containsKey("platform"))
+
+        XCTAssertMatches(savedEvent[mapKeyPath: "properties"], expectedData)
+    }
+
+    func test_setDeviceAttributes_expectFinalJSONHasCorrectKeysAndValues() {
+        let givenIdentifier = String.random
+        let givenToken = String.random
+        let customAttributes: [String: Any] = [
+            "source": "test",
+            "debugMode": true
+        ]
+        let expectedData = deviceInfoMock.configureWithMockData().mergeWith(customAttributes)
+
+        customerIO.identify(identifier: givenIdentifier)
+        customerIO.registerDeviceToken(givenToken)
+        customerIO.deviceAttributes = customAttributes
+
+        let allEvents = readTypeFromStorage(key: Storage.Constants.events)
+        let filteredEvents = allEvents.filter { $0.eventType == "track" }
+        XCTAssertEqual(filteredEvents.count, 2, "too many events received")
+
+        guard let savedEvent = filteredEvents.last else {
+            XCTFail("saved event must not be nil")
+            return
+        }
+
+        XCTAssertEqual(savedEvent[keyPath: "event"] as? String, "Device Created or Updated")
+        XCTAssertEqual(savedEvent[keyPath: "userId"] as? String, givenIdentifier)
+        XCTAssertEqual(savedEvent[keyPath: "context.device.token"] as? String, givenToken)
+
+        XCTAssertMatches(
+            savedEvent[mapKeyPath: "properties"],
+            expectedData,
+            withTypeMap: [["debugMode"]: Bool.self]
```

Comment:
`debugMode` what is this for?

Reply:
Just dummy values for testing custom attributes.